compiler/verification/conditions.ml

```diff
@@ -407,10 +439,15 @@ let rec generate_verification_conditions_scopes
                  destructured with a series of let bindings just after. )*);
           }
         in
-        let _, vcs =
+        let _, vcs, asserts =
           generate_verification_conditions_scope_body_expr ctx scope_body_expr
         in
-        vcs
+        let combined_assert =
+          conjunction_exprs asserts
+            (Typed
+               { pos = Pos.no_pos; ty = Marked.mark Pos.no_pos (TLit TBool) })
```

Comment:
There is probably a better way of doing this, but I'm not sure how

Reply:
Looks nice to me actually. The only thing I might be afraid of is that if this big assert expr is duplicated in all Z3 queries then it will also carry asserts with variables that have nothing to do with the current query. Not sure if it's a big problem.

Reply:
Right, I think there is room for improvement and further optimization of the queries. However, in the current state, the asserts should remain small terms, and only be added in the scopes corresponding to the definition. I think this is fine for now, happy to revisit if we see any issue